File: lab6/min_search_cost.py
def min_search_cost(keys, freq):
    n = len(keys)
    dp = [[0] * n for _ in range(n)]
    freg_sum = [[0] * n for _ in range(n)]
    
    for i in range(n):
        freg_sum[i][i] = freq[i]
        for j in range(i+1, n):
            freg_sum[i][j] = freg_sum[i][j - 1] + freq[j]
    
    for length in range(1, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            dp[i][j] = float('inf')
            for r in range(i, j+1):
                cost = (
                    (dp[i][r-1] if r > i else 0) +
                    (dp[r+1][j] if r < j else 0) +
                    freg_sum[i][j]
                )
                dp[i][j] = min(dp[i][j], cost)
    return dp[0][n-1]

keys = [5, 6]
freq = [17, 25]
print("min search cost: ", min_search_cost(keys, freq))
keys = [5, 10, 15]
freq = [10, 30, 20]
print("min search cost: ", min_search_cost(keys, freq))

File: lab6/MinSearchCost.java



public class MinSearchCost{
    public static int findMinCost(int[] keys, int[] freq) {
        int n = keys.length;
        int[][] dp = new int[n][n];
        int[] freqSum = new int[n + 1];

        for(int i=1; i<=n; i++) freqSum[i] = freqSum[i-1] + freq[i-1];

        for(int len = 1; len <=n; len++) for(int i =0; i<= n - len; i++){
            int j = i + len - 1;
            dp[i][j] = Integer.MAX_VALUE;

            for(int r =i; r<=j; r++){
                int left = (r > i) ? dp[i][r-1] : 0;
                int right = (r < j) ? dp[r+1][j] : 0;
                int cost = left + right + (freqSum[j+1] - freqSum[i]);
                dp[i][j] = Math.min(dp[i][j], cost);
            }
        }
        return dp[0][n-1];
    }
    public static void main(String[] args) {
        int[] keys = {5,6};
        int[] freq = {17, 25};

        System.out.println("Min cost: " + findMinCost(keys, freq));
    }
}